WalletWasabi/BitcoinCore/P2pNode.cs

```diff
@@ -126,65 +127,80 @@ private void P2pNode_StateChanged(Node node, NodeState oldState)
 			}
 		}
 
+		private bool _disposed = false;
+
 		public async Task DisposeAsync()
 		{
-			Stop?.Cancel();
+			if (_disposed)
+			{
+				return;
+			}
+
+			Stop.Cancel();
 			await ReconnectorTask.ConfigureAwait(false);
-			Disconnect();
-			Stop?.Dispose();
-			Stop = null;
+			await DisconnectAsync(CancellationToken.None).ConfigureAwait(false);
+			Stop.Dispose();
+
+			_disposed = true;
```

Comment:
I'm not sure whether `P2pNode` class is supposed to be thread-safe or not. 

Anyway, I would put this line just after line 137.

Reply:
ACK. I am mimicking the dispose pattern here.